### PR TITLE
Laying foundations for a major client overhaul

### PR DESCRIPTION
**Pull request by jhamon**
## Problem

This PR is all about making some foundational changes to pave the way for new client development.

Apologies for the large diff; a lot of things were exploratory at first and once I had things shaping up into something useful it was really difficult to split it up into smaller PRs. None of this is perfect and I'll continue to iterate on it, but for now I need to land some this initial work so that I can take next steps in small focused chunks.

### Code Reorg

- Create a `src/v0` folder to hold legacy code that used to be directly under `src`
    - Move `src/index.ts` > `src/v0/index.ts`
    - Move `src/utils.ts` > `src/v0/utils.ts`
- Create a new `src/index.ts` whose sole responsibility is to declare module exports and adjust paths to find legacy code in its new location. For reasons of backwards compatibility we will continue to export existing exports as they are while we work on a new major version of the node client.
- Move `tests/index.test.ts` > `tests/legacy-integration/v0.test.ts`. In the past these were just called "tests" but they are fully dependent on a prod environment so I have placed them in a folder name that reflects the integration nature of these tests. These can now be run locally with `npm run tests:legacy` although they've been broken for quite a while now so I've temporarily dropped them from CI. I have a separate ticket to fix these tests.

### New export `Pinecone` added to hold new client code

- I introduce a new top-level export, `Pinecone`, which will be where we namespace new development. (The legacy client export is called `PineconeClient`).
- So far this new `async Pinecone.createClient()` method doesn't do anything except validate configuration. Sounds simple, but there's enough complexity it needed a lot of test cases to capture the major edge cases.

### Added a new dependency on `@types/web`

We want to better leverage Typescript in the revised client than we have in the past. This dependency is [developed by Microsoft](https://github.com/microsoft/TypeScript-DOM-lib-generator) and licensed under the permissive [Apache 2 License](https://github.com/microsoft/TypeScript-DOM-lib-generator/blob/main/LICENSE.txt). It is needed to access type definitions for `fetch`.

### Automated Testing

- Introduce unit tests for new client code under `src/__tests__`. Going forward, unit tests will live alongside code under the `src/` directory. These can be run with `npm run test:unit`
- Add a small number of integration tests for the new client initialization under `tests/integration`. These are run with `npm run test:integration`. These are trivial tests at this point because I haven't implemented any of the control plane, but having some starter tests allowed me to setup the CI checks I'll want for future PRs.
- Add integration and unit testing steps to the PR github workflow

### Manual Testing

I setup a script to quickly configure clients in the node REPL for manual testing because that's something I'm doing often to uncover important edge cases.

To use it, set `PINECONE_API_KEY` and `PINECONE_ENVIRONMENT` credentials in a `.env` file, then run `npm run repl`

## Type of Change

- [x] New feature (non-breaking change which adds functionality)

## Test Plan

```
npm run build
npm run test:unit
npm run test:integration
```

And manual testing

```
npm run repl
```

### CODE REVIEW

**Review by micmmakarov — APPROVED** [review]
Looks great!

File: README.md [hunk 1]
```diff
@@ -4,7 +4,6 @@ This is the Node.js client for Pinecone, written in Typescript. It is a wrapper
 
 [![Tests](https://github.com/pinecone-io/pinecone-ts-client/actions/workflows/PR.yml/badge.svg?branch=main)](https://github.com/pinecone-io/pinecone-ts-client/actions/workflows/PR.yml)
 [![Maintenance](https://img.shields.io/badge/Maintained%3F-yes-green.svg)](https://GitHub.com/pinecone-io/pinecone-ts-client/graphs/commit-activity)
-[![Npm package version](https://badgen.net/npm/v/pinecone-ts-client)](https://npmjs.com/package/@pinecone-database/pinecone)
```
Comment (micmmakarov) [inline README.md:7]: lol shrug
